lib/src/parser.dart

```diff
@@ -300,3 +310,47 @@ class Parser {
         '"$value" is not an allowed value for option "${option.name}".');
   }
 }
+
+/// A `List` which uses another `List` as its underlying data structure, but
+/// with the elements in reverse.
+///
+/// The makes `removeAt(0)` much faster, which is critical for parser
+/// performance.
+///
+/// It would be better to use a `Queue`, but `List` was already part of the
+/// public API of `Parser` when the performance issue was noticed.
+class _ReverseList extends Object
+    with ListMixin<String>
+    implements List<String> {
+  final List<String> _list;
+
+  _ReverseList.of(List<String> list) : _list = list.reversed.toList();
+
+  @override
+  int get length => _list.length;
+
+  @override
+  set length(int length) {
+    // Elements are reversed, so if we need to drop elements, drop them from
+    // the end; there's no need to support increasing length.
+    if (length < _list.length) {
+      _list.removeRange(0, _list.length - length);
+    } else if (length == _list.length) {
+      // Nothing to do.
+    } else {
+      throw UnsupportedError("Can't increase length.");
```

Comment:
We could support this by inserting nulls at the front... if we are exposing this list to users we should probably make it work as much like a normal list as possible?